### DISCUSSION TITLE
Model fine-tuning - Train a specific class of the pretrained YOLOv8 model without compromising information related to other classes

### DISCUSSION THREAD

**Discussion by Masoudvahid** [ultralytics/ultralytics#6849] [Q&A]
I am using the pretrained yolov8 model trained on coco dataset, but there is a case that the model does not perform ideally on a specific class.
So I created the desired dataset for this specific class. Then I started to train on the new dataset (300 images) and by freezing the first 10 layers (as far as I understood first 10 layers are backbone). 
The result of training is acceptable, but now I am missing the all other 78 classes. 
My question is how can I fine-tune the model without loosing other classes and only training on very small number of images (without training on coco dataset again).

The command which I used is as follows:
```yolo detect train imgsz=640 batch=16 epochs=50 data="dataset.yaml" model=yolov8x.pt freeze=10```

where dataset.yaml is the path to the new dataset and all default coco classes.

It worthsnoting that when I check the number of classes on the output later of the model, I see all classes, but the model predicts none of them.

And lastly, I also freezed all 22 layers, and retrained. Just in case of making sure if any of the classes would resist, but all the classes were predicting nothing.

**Comment by pderrenger** [comment]
@Masoudvahid when fine-tuning a pretrained model on a new dataset for a specific class, it's important to ensure that the new dataset includes examples of all the classes you want the model to recognize. If you train only on the new class, the model will likely forget the other classes, a phenomenon known as catastrophic forgetting.

To fine-tune on a specific class without losing performance on the other classes, you should include examples of the other classes in your training dataset. This can be a mix of new images for the class of interest and a subset of the original dataset images for the other classes. This way, the model maintains its ability to recognize the other classes while improving on the new class.

Here's a general approach to fine-tuning your model:

1. **Create a Mixed Dataset**: Combine your new images for the specific class with a representative subset of the original COCO dataset that includes all classes.

2. **Use a Lower Learning Rate**: When fine-tuning, use a lower learning rate to make smaller updates to the weights, which helps in retaining the previously learned information.

3. **Selective Freezing**: Instead of freezing the first few layers, you might consider freezing only the layers that are less likely to need retraining, such as the early convolutional layers. The later layers, which are more class-specific, can be fine-tuned.

4. **Class Weights**: Adjust the loss function to give more weight to the new class during training, which can help the model learn from the smaller number of examples without forgetting the other classes.

5. **Regularization Techniques**: Use techniques like dropout or data augmentation to prevent overfitting to the new class.

6. **Monitor Validation Performance**: Ensure you have a validation set that includes all classes. Monitor the performance on this set to ensure that the model retains its ability to detect all classes.

Here's an example command that incorporates some of these principles (note that you'll need to adjust the learning rate and other hyperparameters according to your specific needs):

```python
from ultralytics import YOLO

# Load the pretrained model
model = YOLO('yolov8x.pt')

# Fine-tune the model on the mixed dataset with a lower learning rate
results = model.train(data='mixed_dataset.yaml', epochs=50, lr0=0.001, freeze=10)
```

Remember to create a `mixed_dataset.yaml` that includes paths to both the new class images and a subset of the original dataset. Adjust the learning rate (`lr0`) as needed. The `freeze` parameter should be set based on the layers you want to freeze, if any.

Keep in mind that fine-tuning is a delicate process, and it may take some experimentation to get the balance right between retaining old knowledge and learning new information. 🧪🔍

**Reply by xihri412** [reply]
@pderrenger For the mixed dataset, should I label the old classes in the new dataset and label the new classes in the COCO dataset?

**Reply by pderrenger** [reply]
Hi @xihri412! Yes, for effective fine-tuning, you should label the old classes in your new dataset. This helps maintain the model's ability to recognize these classes. There's no need to label new classes in the original COCO dataset unless those new classes are also present and you want to improve detection for them as well. Essentially, ensure that your training data is representative of all classes you want the model to detect. Happy fine-tuning! 😊

**Reply by APeFerreira** [reply]
@pderrenger how can I adjust the class loss weights when adding a new class?

**Comment by MlLearnerAkash** [comment]
I am also looking for the solution.
Is there any update?

**Reply by glenn-jocher** [reply]
@MlLearnerAkash for fine-tuning a specific class without losing others, consider using a smaller learning rate and ensuring your dataset includes annotations for all classes, even if they're not present. This helps retain the model's knowledge of other classes.

**Reply by rtitton** [reply]
Hello, how many instances/images from COCO should I include in my dataset (along with images of my new class) to achieve a fine-tuned model with performance metrics similar to those of the pretrained model? I've conducted some experiments and noticed that my fine-tuned models are significantly less effective at detecting COCO classes (for example, the person class).

**Reply by glenn-jocher** [reply]
To maintain performance on COCO classes while fine-tuning on a new class, include at least 1000-1500 images per retained COCO class in your dataset. Ensure annotations for all classes are accurate and consistent. Fine-tuning with a balanced dataset and a low learning rate can help preserve the pretrained knowledge.

**Reply by rtitton** [reply]
Thank you, I will experiment. 
As you didn't mention it I guess playing with freeze parameter is not an option in this case.
Finally, can you suggest any tool that helps in subsetting COCO?

**Reply by glenn-jocher** [reply]
You're correct; the `freeze` parameter wouldn't help retain COCO knowledge in this case. For subsetting COCO, consider using tools like [FiftyOne](https://github.com/voxel51/fiftyone) or [pycocotools](https://github.com/cocodataset/cocoapi), which allow efficient dataset manipulation and subsetting.

**Reply by rtitton** [reply]
Is there any particular action to take when dealing with classes that can easily be confused? for example "pedestrian" and "person-riding-electric-scooter"?
Thank you for any help.

**Reply by glenn-jocher** [reply]
When dealing with similar classes, ensure your dataset annotations are precise and consistent, and consider increasing the number of training samples for these classes. You can also adjust the loss function or class weights to emphasize better differentiation during training.
